hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/AbstractClientInternalCacheProxy.java

```diff
@@ -912,9 +925,66 @@ protected ICompletableFuture createCompletedFuture(Object value) {
                 clientContext.getExecutionService().getAsyncExecutor());
     }
 
-    private final class NearCacheRepairingHandlerAdapter extends CacheAddInvalidationListenerCodec.AbstractEventHandler
+
+    /**
+     * Needed to deal with client compatibility issues.
+     * Eventual consistency for near cache can be used with server versions >= 3.8.
+     * Other connected server versions must use {@link LegacyNearCacheEventHandler}
+     */
+    private final class ConnectedServerVersionAwareNearCacheEventHandler implements EventHandler<ClientMessage> {
+
+        private final RepairableNearCacheEventHandler repairingHandler = new RepairableNearCacheEventHandler();
+        private final LegacyNearCacheEventHandler legacyHandler = new LegacyNearCacheEventHandler();
+        private volatile boolean supportsRepairableNearCache;
+
+        @Override
+        public void beforeListenerRegister() {
+            legacyHandler.beforeListenerRegister();
+            repairingHandler.beforeListenerRegister();
+        }
+
+        @Override
+        public void onListenerRegister() {
+            supportsRepairableNearCache = supportsRepairableNearCache();
```

Comment:
this handles listener re-registration in case of server version change. For ex. 3.8 client connected 3.7 server and later server version is changed to 3.8 or vice versa. In these cases appropriate listener should be registered.